Extend PublicSuffix List with convenience method to access URL domain without the public suffix

brentfulgham opened the pull request:
#### 72e7d13f0b054ebe4abf9e8f7a5bdfc404f89a53
<pre>
Extend PublicSuffix List with convenience method to access URL domain without the public suffix
<a href="https://bugs.webkit.org/show_bug.cgi?id=284650">https://bugs.webkit.org/show_bug.cgi?id=284650</a>
&lt;<a href="https://rdar.apple.com/problem/141452522">rdar://problem/141452522</a>&gt;

Reviewed by Matthew Finkel.

Add two new convenience methods to the PublicSuffixStore that allow a StringView to be passed, returning a String that is the passed domain with any Public Suffix removed.

This will be used for future Quirk pattern matching.

Tested by new API tests.

* Source/WebCore/platform/PublicSuffixStore.cpp:
(WebCore::PublicSuffixStore::topPrivatelyControlledDomainWithoutPublicSuffix const):
(WebCore::PublicSuffixStore::domainWithoutPublicSuffix const):
* Source/WebCore/platform/PublicSuffixStore.h:
* Tools/TestWebKitAPI/Tests/WebCore/PublicSuffix.cpp:
(TestWebKitAPI::TEST_F(PublicSuffix, topPrivatelyControlledDomainWithoutPublicSuffix)):

Canonical link: <a href="https://commits.webkit.org/287886@main">https://commits.webkit.org/287886@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1663f1ccb6fd6288bfd2dbf9bbaa1c2b8f69acf3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80996 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/521 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/34939 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/85525 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31982 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/83106 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/538 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/8322 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/63239 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21004 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/84065 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/73718 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/43537 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/27876 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/30440 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/28427 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86960 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/8226 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5807 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/71542 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/8403 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/69551 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70777 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14813 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13735 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12586 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/8187 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/13710 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/8024 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/11544 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/9832 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sysrqb commented on Source/WebCore/platform/PublicSuffixStore.cpp:
```diff
@@ -96,4 +96,33 @@ String PublicSuffixStore::topPrivatelyControlledDomain(StringView host) const
+        if (isPublicSuffix(candidate)) {
+            if (separatorPosition > 0 && domain.length() > separatorPosition) {
+                auto domainWithoutSuffix = domain.substring(0, separatorPosition);
+                auto foo = domainWithoutSuffix.toString();
```
Seems like this can be deleted.

brentfulgham replied:
Oh yes -- stupid debugging code I forgot about. Will fix.

brentfulgham committed:
Extend PublicSuffix List with convenience method to access URL domain without the public suffix
https://bugs.webkit.org/show_bug.cgi?id=284650
<rdar://problem/141452522>

Reviewed by Matthew Finkel.

Add two new convenience methods to the PublicSuffixStore that allow a StringView to be passed, returning a String that is the passed domain with any Public Suffix removed.

This will be used for future Quirk pattern matching.

Tested by new API tests.

* Source/WebCore/platform/PublicSuffixStore.cpp:
(WebCore::PublicSuffixStore::topPrivatelyControlledDomainWithoutPublicSuffix const):
(WebCore::PublicSuffixStore::domainWithoutPublicSuffix const):
* Source/WebCore/platform/PublicSuffixStore.h:
* Tools/TestWebKitAPI/Tests/WebCore/PublicSuffix.cpp:
(TestWebKitAPI::TEST_F(PublicSuffix, topPrivatelyControlledDomainWithoutPublicSuffix)):

Canonical link: https://commits.webkit.org/287886@main
